src/lib.rs

```diff
@@ -180,60 +197,95 @@ pub async fn start_lemmy_server() -> Result<(), LemmyError> {
-          lemmy_apub::http::routes::config(cfg);
-          webfinger::config(cfg);
-        }
-      })
-      .configure(feeds::config)
-      .configure(|cfg| images::config(cfg, pictrs_client.clone(), rate_limit_cell))
-      .configure(nodeinfo::config)
-  })
-  .bind((settings_bind.bind, settings_bind.port))?
-  .run()
-  .await?;
-
-  // Wait for outgoing apub sends to complete
-  ActivityChannel::close(outgoing_activities_task).await?;
+
+  let server = if args.http_server {
+    let federation_config = federation_config.clone();
+    // Create Http server with websocket support
+    let server = HttpServer::new(move || {
+      let cors_origin = env::var("LEMMY_CORS_ORIGIN");
+      let cors_config = match (cors_origin, cfg!(debug_assertions)) {
+        (Ok(origin), false) => Cors::default()
+          .allowed_origin(&origin)
+          .allowed_origin(&settings.get_protocol_and_hostname()),
+        _ => Cors::default()
+          .allow_any_origin()
+          .allow_any_method()
+          .allow_any_header()
+          .expose_any_header()
+          .max_age(3600),
+      };
+
+      let app = App::new()
+        .wrap(middleware::Logger::new(
+          // This is the default log format save for the usage of %{r}a over %a to guarantee to record the client's (forwarded) IP and not the last peer address, since the latter is frequently just a reverse proxy
+          "%{r}a '%r' %s %b '%{Referer}i' '%{User-Agent}i' %T",
+        ))
+        .wrap(middleware::Compress::default())
+        .wrap(cors_config)
+        .wrap(TracingLogger::<QuieterRootSpanBuilder>::new())
+        .wrap(ErrorHandlers::new().default_handler(jsonify_plain_text_errors))
+        .app_data(Data::new(context.clone()))
+        .app_data(Data::new(rate_limit_cell.clone()))
+        .wrap(FederationMiddleware::new(federation_config.clone()))
+        .wrap(SessionMiddleware::new(context.clone()));
+
+      #[cfg(feature = "prometheus-metrics")]
+      let app = app.wrap(prom_api_metrics.clone());
+
+      // The routes
+      app
+        .configure(|cfg| api_routes_http::config(cfg, rate_limit_cell))
+        .configure(|cfg| {
+          if federation_enabled {
+            lemmy_apub::http::routes::config(cfg);
+            webfinger::config(cfg);
+          }
+        })
+        .configure(feeds::config)
+        .configure(|cfg| images::config(cfg, pictrs_client.clone(), rate_limit_cell))
+        .configure(nodeinfo::config)
+    })
+    .disable_signals()
+    .bind((settings_bind.bind, settings_bind.port))?
+    .run();
+    let handle = server.handle();
+    tokio::task::spawn(server);
+    Some(handle)
+  } else {
+    None
+  };
+  let federate = if args.federate_activities {
+    Some(start_stop_federation_workers_cancellable(
+      Opts {
+        process_index: args.federate_process_index,
+        process_count: args.federate_process_count,
+      },
+      pool.clone(),
+      federation_config.clone(),
+    ))
+  } else {
+    None
+  };
```

Comment:
```suggestion
  let federate = args.federate_activities.then(|| {
    start_stop_federation_workers_cancellable(
      Opts {
        process_index: args.federate_process_index,
        process_count: args.federate_process_count,
      },
      pool.clone(),
      federation_config.clone(),
    )
  });
```

Use `.then` instead if-else to automatically handle the none case.